executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,751 @@
+	}
+	req.Reset()
+	task := e.getFinishedTask(ctx)
+	for task != nil {
+		task.wgSetResultsSize.Wait()
+		select {
+		case chk := <-task.results:
+			req.Append(chk, 0, chk.NumRows())
+			return nil
+		case err := <-task.doneErr:
+			task.done = true
+			if err != nil {
+				return err
+			}
+			task = e.getFinishedTask(ctx)
+		case <-ctx.Done():
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (e *IndexLookUpMergeJoin) getFinishedTask(ctx context.Context) *lookUpMergeJoinTask {
+	task := e.task
+	if task != nil {
+		task.wgSetResultsSize.Wait()
+		if !task.done || len(task.results) > 0 {
+			return task
+		}
+	}
+
+	select {
+	case task = <-e.resultCh:
+	case <-ctx.Done():
+		return nil
+	}
+	if task == nil {
+		return nil
+	}
+
+	e.task = task
+	return task
+}
+
+func (omw *outerMergeWorker) run(ctx context.Context, wg *sync.WaitGroup) {
+	defer func() {
+		if r := recover(); r != nil {
+			err := errors.Errorf("%v", r)
+			logutil.Logger(context.Background()).Error("outerMergeWorker panic in the recoverable goroutine", zap.Error(err))
+			task := &lookUpMergeJoinTask{
+				results: make(chan *chunk.Chunk, 1),
+				doneErr: make(chan error, 1),
+				done:    true,
+			}
+			task.doneErr <- err
+			omw.pushToChan(ctx, task, omw.resultCh)
+		}
+		close(omw.resultCh)
+		close(omw.innerCh)
+		wg.Done()
+	}()
+	for {
+		task, err := omw.buildTask(ctx)
+		if err != nil {
+			task.doneErr <- err
+			task.wgSetResultsSize.Done()
+			omw.pushToChan(ctx, task, omw.resultCh)
+			return
+		}
+		if task == nil {
+			return
+		}
+
+		if finished := omw.pushToChan(ctx, task, omw.innerCh); finished {
+			return
+		}
+
+		if finished := omw.pushToChan(ctx, task, omw.resultCh); finished {
+			return
+		}
+	}
+}
+
+func (omw *outerMergeWorker) pushToChan(ctx context.Context, task *lookUpMergeJoinTask, dst chan<- *lookUpMergeJoinTask) bool {
+	select {
+	case <-ctx.Done():
+		return true
+	case dst <- task:
+	}
+	return false
+}
+
+// buildTask builds a lookUpMergeJoinTask and read outer rows.
+// When err is not nil, task must not be nil to send the error to the main thread via task
+func (omw *outerMergeWorker) buildTask(ctx context.Context) (*lookUpMergeJoinTask, error) {
+	newFirstChunk(omw.executor)
+	task := &lookUpMergeJoinTask{
+		doneErr:     make(chan error, 1),
+		outerResult: newFirstChunk(omw.executor),
+	}
+	task.wgSetResultsSize.Add(1)
+	task.memTracker = memory.NewTracker(stringutil.MemoizeStr(func() string { return fmt.Sprintf("lookup join task %p", task) }), -1)
+	task.memTracker.AttachTo(omw.parentMemTracker)
+
+	omw.increaseBatchSize()
+	if omw.lookup.isOuterJoin { // if is outerJoin, push the requiredRows down
+		requiredRows := int(atomic.LoadInt64(&omw.lookup.requiredRows))
+		task.outerResult.SetRequiredRows(requiredRows, omw.maxBatchSize)
+	} else {
+		task.outerResult.SetRequiredRows(omw.batchSize, omw.maxBatchSize)
+	}
+
+	task.memTracker.Consume(task.outerResult.MemoryUsage())
+	for !task.outerResult.IsFull() {
```

Comment:
It may be ok to wrap the `omw.executorChk` into a task. The child would promise the fullness of omw.executorChk.